### PR TITLE
Unittest added for asserting the chemical shifts

### PR DESCRIPTION
**Pull request by mehradans92**
This unittest asserts the chemical shifts based on the plumed data. The first test checks for the parser function which obtains the indices of the used chemical shifts. The other tests evaluate the functionality of validity_check func for different inputs. Function validity_check compares data and raises a ValueError for those used chemical shifts that are equal to 1.

### CODE REVIEW

**Comment by hgandhi2411** [comment]
@mehradans92 your code ( functions like parser and validity check) should be defined in the peptidesim package and your test file should import and call that function. You should not be defining your functions in the test file.

**Comment by whitead** [comment]
Take a look at this example PR @mehradans92 [github.com/ur-whitelab/hoomd-tf/pull/88](https://github.com/ur-whitelab/hoomd-tf/pull/88). Notice how there is code for a test and code that modifies the code base. That's what you want to see.

**Comment by whitead** [comment]
Per our discussion:

- [ ] Move test code to `test_utils.py`
- [ ] Move parser code to `utils.py`
- [ ] Work with Dilnoza to add to `part1.py`

**Comment by mehradans92** [comment]
The unittest has been added as a new file (test_utils.py) in /package/tests. The parser function and the cs_validity function have been added to utilities.py. The fuction cs_validity two inputs which are the path the plumed data and cs_shifts data and asserts if the chemical shifts are valid. See below:
`cs_validity("path to plumed data (followed by the name of the plumedfile.dat)","path to the chemical shift dat files")`.
This function can be simply added to maybe part1.py.
Please re-review @whitead @dilnoza92

File: package/peptidesim/utilities.py [hunk 1]
```diff
@@ -39,3 +39,72 @@ def in_directory(file, directory, allow_symlink=False):
...
+        if df.values[Index][1] == 1:
+            Inval_index.append(Index+1)
+            count += 1
+    return (count, Inval_index, Shif_index)
+
+
+def cs_validity(plumed_dat, exec_dir):
+    """plumed_dat = The path to the plumed file (include the plumed data filename and extension)
+       exec_dir = The path to the cs_shifts data files)"""
+    file_plumed = open(plumed_dat, 'r')
```
Comment (whitead) [inline package/peptidesim/utilities.py]: Use the `with open(...)` syntax for working with python files.

File: package/peptidesim/utilities.py [hunk 2]
```diff
@@ -39,3 +39,72 @@ def in_directory(file, directory, allow_symlink=False):
...
+    shifts_plumed = parser(file_plumed)
+    file_plumed.close()
+    keys = ['CAshifts', 'CBshifts', 'Cshifts',
+            'HAshifts', 'Hshifts', 'Nshifts']
+    shifts_plumed_dict = dict.fromkeys(keys)
+    dat_files = [f for f in os.listdir(exec_dir) if f.endswith('.dat')]
+    for i in range(len(shifts_plumed)):
+        shifts_dat = exec_dir + dat_files[i]
+        df_shifts_dat = pd.read_csv(shifts_dat, sep=" ", header=None)
```
Comment (whitead) [inline package/peptidesim/utilities.py]: Use only single quotes

Comment (whitead) [inline package/peptidesim/utilities.py]: Is there anyway to use the built-in python `csv` module or must we add pandas as a dependency?

File: package/peptidesim/utilities.py [hunk 3]
```diff
@@ -39,3 +39,72 @@ def in_directory(file, directory, allow_symlink=False):
...
+        shifts_plumed_dict[keys[i]] = shifts_plumed[i]
+        if shifts_plumed[i] != []:
+            count, Invalid_index, Shifts_index = validity_check(
+                df_shifts_dat, shifts_plumed_dict[keys[i]])
+            if count != 0:
+                raise ValueError('Simulation uses data {} for {}. Invalid EDS chemical shift found for data {}!'.format(
+                    Shifts_index, keys[i], Invalid_index))
+    print('All chemical shifts are valid'.)
```
Comment (whitead) [inline package/peptidesim/utilities.py]: Can you just return a `True`? We don't use `print` in the peptidesim module, instead using a log file. The `print` statements end up in the slurm file.

File: package/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import numpy as np
```
Comment (whitead) [inline package/tests/test_utils.py]: Let's make the names match. Maybe rename to `test_utilities.py`

File: package/peptidesim/utilities.py (hunk 3 above)
Comment (whitead) [reply package/peptidesim/utilities.py]: Also the period is outside the string, giving me a syntax error.

File: package/peptidesim/utilities.py (hunk 2 above)
Comment (mehradans92) [reply package/peptidesim/utilities.py]: Thanks for your comments @whitead . I think I accounted for all the requested changes. I guess you could use python default csv but I used Pandas.

**Comment by whitead** [comment]
I get this error when running the tests:

```
tests/test_utils.py:14
  /home/whitelab/scratch/peptidesim/package/tests/test_utils.py:14: DeprecationWarning: invalid escape sequence \.
    txt_plumed = ('PRINT ARG=(cs\.cb_.*),(cs\.ca_.*),(cs\.hn_.*),(cs\.ha_.*),(cs\.nh_.*),(cs\.co_.*) FILE=WT_PTE_CS_shifts STRIDE=250\n' +
```

Also, please make sure you only have single quotes in your python source file once more.

**Comment by mehradans92** [comment]
Thanks for your comments @whitead . Issue fixed.

**Comment by hgandhi2411** [comment]
I get the same error as @whitead when I run the tests. 
Also, can you make the file name `test_utilities.py` for consistency.

**Comment by whitead** [comment]
I see the following error:

```
___________________________________________________________________________________________ Test.test_cs_shifts_example ___________________________________________________________________________________________

self = <tests.test_utilities.Test testMethod=test_cs_shifts_example>

    def test_cs_shifts_example(self):
        txt_plumed = ('PRINT ARG=(cs\.cb_.*),(cs\.ca_.*),(cs\.hn_.*),(cs\.ha_.*),(cs\.nh_.*),(cs\.co_.*) FILE=WT_PTE_CS_shifts STRIDE=250\n' +
                      'eds: EDS ARG=cs.hn_1,cs.hn_2,cs.hn_3,cs.hn_10,cs.nh_14,cs.nh_15,cs.nh_16,cs.nh_22,cs.cb_1,cs.cb_11,cs.cb_16,cs.ha_1,cs.ha_10,cs.ca_1,cs.ca_2,cs.ca_5,cs.c_3,cs.c_5,cs.c_7 CENTER_ARG=cs.exphn_1,cs.exphn_2,cs.exphn_3,cs.exphn_4,cs.exphn_6,cs.exphn_7,cs.exphn_8,cs.exphn_9,cs.exphn_10,cs.exphn_13,cs.exphn_14,cs.exphn_15,cs.exphn_16,cs.exphn_18,cs.exphn_19,cs.exphn_20,cs.exphn_21,cs.exphn_22,cs.expnh_1,cs.expnh_2,cs.expnh_3,cs.expnh_4,cs.expnh_6,cs.expnh_7,cs.expnh_8,cs.expnh_9,cs.expnh_10,cs.expnh_13,cs.expnh_14,cs.expnh_15,cs.expnh_16,cs.expnh_18,cs.expnh_19,cs.expnh_20,cs.expnh_21,cs.expnh_22 MULTI_PROP=0.4 RANGE=0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1,0.1 PERIOD=250 TEMP=@replicas:{278.0,284.825679594,291.818948762,298.9839223,306.324816031,313.84594929,321.551747461,329.446744586,337.535586031,345.823031217,354.313956423,363.013357653,371.926353579,381.05818855,390.414235678,400.0} OUT_RESTART=eds_bias_restart_correct3.dat\n' +
                      'PRINT ARG=eds.bias,eds.force2 STRIDE=200 FILE=eds.dat')
        plumed_dat = 'Plumed.dat'
        with open(plumed_dat, 'w') as f1:
            f1.write(txt_plumed)
        keys = ['CAshifts', 'CBshifts', 'Cshifts',
                'HAshifts', 'Hshifts', 'Nshifts']
        df_ca = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                    '19', '20', '21', '22', '23', '#24'], '1': [1.5, 11.5, 1, 1, 1.5, 1.5, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]})
        df_cb = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                    '19', '20', '21', '22', '23', '#24'], '1': [1.5, 1, 1.5, 1, 1.5, 1, 1.5, 1, 1, 1, 1.2, 1, 1, 1, 1, 1.2, 1, 1, 1, 1, 1, 1, 1, 1]})
        df_c = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                   '19', '20', '21', '22', '23', '#24'], '1': [1, 1, 1.5, 1.5, 1.2, 1, 1.6, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]})
        df_ha = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                    '19', '20', '21', '22', '23', '#24'], '1': [1.5, 1.2, 1, 1, 1, 1, 1, 1, 1, 1.3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]})
        df_h = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                   '19', '20', '21', '22', '23', '#24'], '1': [1.5, 1.5, 8.51, 8.39, 1, 8.5, 8.46, 8.52, 8.62, 8.02, 8.44, 8.14, 1, 1, 8.51, 8.39, 8.25, 1, 8.46, 8.52, 8.62, 8.02, 8.44, 8.14]})
        df_n = pd.DataFrame({'0': ['#1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '#12', '#13', '14', '15', '16', '17', '18',
                                   '19', '20', '21', '22', '23', '#24'], '1': [1, 1, 120.7, 122.2, 123.8, 121.4, 122, 116.9, 111, 120.4, 122.7, 121.5, 1, 10, 120.7, 122.2, 123.8, 121.4, 122.2, 123.8, 1111, 120.4, 122.7, 121.5]})
        df = [df_ca, df_cb, df_c, df_ha, df_h, df_n]
        for i in range(len(keys)):
            shifts_dat = keys[i] + '.dat'
            df[i].to_csv(shifts_dat, index=False,
                         index_label=False, sep=' ', header=False)
    
        with open(plumed_dat, 'r') as file_plumed:
            shifts_plumed = utilities.parser(file_plumed)
        os.remove(plumed_dat)
        exec_dir = os.getcwd()
        dat_files = [f for f in os.listdir(exec_dir) if f.endswith('.dat')]
        shifts_plumed_dict = dict.fromkeys(keys)
        for i in range(len(shifts_plumed)):
            shifts_dat = exec_dir + '//' + dat_files[i]
>           df_shifts_dat = pd.read_csv(shifts_dat, sep=' ', header=None)

tests/test_utilities.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py37/lib/python3.7/site-packages/pandas/io/parsers.py:676: in parser_f
    return _read(filepath_or_buffer, kwds)
.tox/py37/lib/python3.7/site-packages/pandas/io/parsers.py:454: in _read
    data = parser.read(nrows)
.tox/py37/lib/python3.7/site-packages/pandas/io/parsers.py:1133: in read
    ret = self._engine.read(nrows)
.tox/py37/lib/python3.7/site-packages/pandas/io/parsers.py:2037: in read
    data = self._reader.read(nrows)
pandas/_libs/parsers.pyx:859: in pandas._libs.parsers.TextReader.read
    ???
pandas/_libs/parsers.pyx:874: in pandas._libs.parsers.TextReader._read_low_memory
    ???
pandas/_libs/parsers.pyx:928: in pandas._libs.parsers.TextReader._read_rows
    ???
pandas/_libs/parsers.pyx:915: in pandas._libs.parsers.TextReader._tokenize_rows
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pandas.errors.ParserError: Error tokenizing data. C error: Expected 3 fields in line 3, saw 4

pandas/_libs/parsers.pyx:2070: ParserError
```

**Comment by mehradans92** [comment]
It was already in the tests directory before this PR. I believe it makes the folder part of the package.

**Comment by whitead** [comment]
Hi @mehradans92 I'm still seeing the failure I saw above. @hgandhi2411 you said the tests pass for you though?

**Comment by hgandhi2411** [comment]
> Hi @mehradans92 I'm still seeing the failure I saw above. @hgandhi2411 you said the tests pass for you though?

Yes the tests pass for me @whitead

**Comment by whitead** [comment]
Alright, good enough for me.

**Pull request merged**